### DISCUSSION TITLE
Cloudflare - cloudflared - secure tunnel access to Frigate etc

### DISCUSSION THREAD

**Discussion by LaurenceGough** [blakeblackshear/frigate#4247] [Show and tell]
Hi all,

After experiencing the almost week long outage of duckdns for my Home assistant setup I looked into an alternative. I just wanted to share my recent experience with Cloudflare's Zero Trust product. You run a docker image called cloudflared which creates a secure tunnel to Cloudflare's network.

No port forwarding required, no cert renewals needed (depending on how you set it up), it's far easier than other methods to setup and I feel offers great security. You can setup ACLs, groups, users, WAF, all kinds of rules and link it with many MFA solutions. 

I believe there is also an addin for Homeassistant OS etc - but I use the core container image now.

I had a domain already so I renewed it with them to keep it all together which makes it even easier and they only charge the cost price for domains!!

You can then securely access your Frigate, Home assistant and any other applications you have seamlessly from the local network and externally.

More info here:

https://www.cloudflare.com/en-gb/products/tunnel/
https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/

It's free too for personal use.

I hope this helps someone.

**Comment by NickM-27** [comment]
I used cloudflare ddns and a few months ago moved to this and it has been awesome.

**Comment by Cold-Lemonade** [comment]
I run ddwrt on my router with Wireguard server. On my iPhone and laptop, I run Wireguard client. This lets me create a secure, encrypted tunnel between my devices and my home network when I'm away. With the Wireguard tunnel in place, I can access all the resources on my home network as if I were there, including Frigate and Home Assistant

**Comment by jeeftor** [comment]
How did you get your web sockets working?

**Reply by NickM-27** [reply]
what error are you seeing? Websockets work fine for me

**Reply by jeeftor** [reply]
> WebSocket connection to 'wss://dvr.MYDOMAIN.net/ws' failed: The network connection was lost.

**Reply by jeeftor** [reply]
The UI opens fine -> but can't seem to get any websocket traffic flowing at all

**Reply by incith** [reply]
This is also my problem.  Live feeds just don't work.  I am using Apache to reverse proxy the traffic.  Even connecting directly doesn't work.  No errors on nginx logs.

Googling lead me to understand wireguard doesn't tunnel websocket traffic.

Edit: I know Apache is working because it works fine on LAN. Used config from frigate docs for proxy settings.

**Reply by jeeftor** [reply]
Odd last night -> it just started working randomly ... 🤷️

**Reply by incith** [reply]
Huh.

Well, I've no idea.  The debug stream works but not live.  Only on local wifi/local network does the live stream seem to work fine.  I was going to try an ipsec VPN in a while instead of wireguard.

**Reply by aneisch** [reply]
Can you paste your config?

**Reply by incith** [reply]
Certainly! --

**Apache vhost configuration:**
```<VirtualHost *:80>
        ServerName nvr.my.lan
        ServerAlias nvr
        Include sites-available/proxies.include
        ErrorLog /var/log/apache2/nvr_error_log

        Redirect / https://nvr.my.lan/
</VirtualHost>

<IfModule ssl_module>
<VirtualHost *:443>
        ServerName nvr.my.lan
        ServerAlias nvr
        Include sites-available/proxies.include
        ErrorLog /var/log/apache2/nvr_ssl_error_log

        SSLProxyEngine On
        SSLProxyCheckPeerCN off
        SSLProxyCheckPeerName off
        ProxyPreserveHost On

        ProxyPass "/"  "http://10.1.5.5:5000/"
        ProxyPassReverse "/"  "http://10.1.5.5:5000/"

        ProxyPass /ws ws://10.1.5.5:5000/ws
        ProxyPassReverse /ws ws://10.1.5.5:5000/ws

        ProxyPass /live/ ws://10.1.5.5:5000/live/
        ProxyPassReverse /live/ ws://10.1.5.5:5000/live/

        RewriteEngine on
        RewriteCond %{HTTP:Upgrade} =websocket [NC]
        RewriteRule /(.*) ws://10.1.5.5:5000/$1 [P,L]
        RewriteCond %{HTTP:Upgrade} !=websocket [NC]
        RewriteRule /(.*) http://10.1.5.5:5000/$1 [P,L]

        <IfModule log_config_module>
                TransferLog /var/log/apache2/nvr_ssl_access_log
        </IfModule>

        SSLEngine on
        SSLProtocol ALL -SSLv2 -SSLv3 -TLSv1
        SSLCipherSuite ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA->
        SSLHonorCipherOrder On

        SSLCertificateFile /certs/my.lan-bundle.crt
        SSLCertificateKeyFile /certs/my.lan.key
        <IfModule log_config_module>
                CustomLog /var/log/apache2/nvr_ssl_request_log \
                        "%t %h %{SSL_PROTOCOL}x %{SSL_CIPHER}x \"%r\" %b"
        </IfModule>
</VirtualHost>
```

**Frigate go2rtc portion:**
```go2rtc:
  webrtc:
    candidates:
      - 10.1.5.5:8555
  streams:
    back:
      - "rtsp://{FRIGATE_RTSP_USERNAME}:{FRIGATE_RTSP_PASSWORD}@10.1.5.101:554/cam/realmonitor?channel=1&subtype=0"
      - "ffmpeg:back#audio=opus"
    hikvision1:
      - "ffmpeg:rtsp://{FRIGATE_RTSP_USERNAME}:{FRIGATE_RTSP_PASSWORD}@10.1.5.102:554/Streaming/Channels/101#video=copy"```

**Reply by aneisch** [reply]
You might need this in go2rtc "candidates"
`      - stun:8555`

**Reply by incith** [reply]
No luck.  As far as the logs are concerned, everything is great.


In fact, what's interesting is wireguard is showing it's receiving MB's of data (when the live view is open, but spinning)...so I'm wondering if there's just not enough bandwidth to stream it over the tunnel at this point.  There's no errors at all in any of the logs.. Apache on proxy server shows good status, etc.

```2023-08-20 17:34:11.399947038  10.1.x.x - - [20/Aug/2023:17:34:11 -0400] "GET /ws HTTP/1.1" 101 23948 "-" "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Mobile Safari/537.36" "192.168.y.y"```

```2023-08-20 17:43:37.045581023  10.1.x.x - - [20/Aug/2023:17:43:37 -0400] "GET /ws HTTP/1.1" 101 2521 "-" "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Mobile Safari/537.36" "10.1.y.y"```

Same exact reply on vpn and off vpn, just only the latter works.  So strange.

Edit: No luck on IPSEC vs. WireGuard either.  Super odd.

**Reply by h2rSys** [reply]
In my case I had to open 8555 port to server on router. Then webrtc works..

**Comment by aneisch** [comment]
Just an FYI but I am pretty sure it is against CloudFlare TOS to deliver video unless using one of the specific paid services they offer. Please correct me if I am misunderstanding. 

https://developers.cloudflare.com/support/more-dashboard-apps/cloudflare-stream/delivering-videos-with-cloudflare/
https://www.cloudflare.com/service-specific-terms-application-services/

**Reply by NickM-27** [reply]
Right, using WebRTC for live view means that the cloudflare tunnel is not used to deliver the video or audio. Watching recordings though will be against the ToS so important to keep that in mind

**Reply by LaurenceGough** [reply]
CloudFlare offer many different products, most aimed at website protection. I couldn't see anything specific regarding the tunnels which I think are a relatively new product from them? Perhaps it hasn't been updated. I average around 5GB per a month with them, I'm sure it's nothing compared to most of their massive customers. Aren't they one of the biggest CDNs as well?

I think it's aimed more at people streaming picture movies and/or TV, massive bandwidth consumers.

If it is against the TACs I guess I and others will need to find an alternative.

To be honest I find watching recordings via it extremely slow, it's the only downside I could say about it. And I know it's not my Frigate instance as it's almost instant locally. My upload bandwidth should be enough too. I wonder if they do any kind of QoS or throttling for it, that would explain it.


https://community.cloudflare.com/t/questions-about-using-streaming-video-through-cloudflare-tunnel/482315

Hmm it doesn't seem crystal clear but it's looking like it might not be allowed, well, I guess I might no longer recommend it 🤣. Probably one of those things where I'd you don't take the pee at all it's OK... I'll look into an alternative...

**Comment by PovilasID** [comment]
I have instance of this running behind mobile CGNAT and it is working very well!
It is only one of only 3 approaches I found that work and this is defiantly the fastest to setup.
It also offers options to setup logins using your own email or adding even your own OIDC provider.

**Comment by dopeytree** [comment]
I would use tailscale instead on the network https://tailscale.com
It's also free for up to 3 users & 20 devices. 

(I've put it on a pfsense router so the whole subnet counts as a device!) Then only the device/s used to access to network remotely count against the free 20 limit.

**Reply by jeeftor** [reply]
But cloudflare you dont need a client. I've never got the SSH/VNC working through cloudflare - I've never got the VNC working through tail scale (web based).

So when coming in from my work phone cloudflare is good

**Reply by dopeytree** [reply]
noVNC works just tested in on my mobile to my unraid machine. Haven't got hostnames working but direct IPs all work on tailscale. You just use the local IP even though its remote https://tailscale.com https://novnc.com

**Comment by incith** [comment]
Well, I just resolved this for myself after some more sleepless nights of wondering WHY this just would not work for me on my setup.

My setup is pfSense and Wireguard for remotely viewing the cameras.

No matter what firewall rules I put in place, ports forwarded, etc etc, the issue seems to have come down to having my frigate machine on a different vlan.  Once I moved it to the main LAN, updated the NAT mapping in pfSense and it worked immediately now.  Super bizarre.

Edit - to clarify, the only thing that wasn't working was webrtc.  It's working now.

Edit2 - OMG...this just lead me to realize what else could be wrong.  My unifi profiles.  I use profiles for switchport configuration and the got dang security profile was blocking all other networks.  Holy crap.

Final edit - I'm actually running frigate in bridge mode now, with just port 8555 forwarded externally on firewall.  I had also changed a few things in frigate config but I doubt they matter (I'm almost certain it was my unifi profiles). In go2rtc section I commented out the listen section and candidates sections.  Go2rtc seems to set them automatically.  I also changed my cameras section to reference rtsp://frigate.lanip instead of localhost.  I doubt either of these two changes contributed to me fixing my problem but thought I'd share anyway.

**Comment by vector-co-uz** [comment]
Hello,
I’m running Frigate behind a Cloudflare Zero Trust Tunnel.
The web UI is fully accessible via a domain and works fine.

However, notifications/events do not work:
no real-time notifications

When I expose Frigate using CloudPub, everything works correctly (events, notifications, real-time updates), so Frigate itself is working as expected.

Setup details:

Frigate is reachable locally and works fine

Cloudflare Tunnel (cloudflared) is used only as a reverse proxy

Issue appears only when accessed through Cloudflare Tunnel

No issue when accessed  via CloudPub


Any guidance or recommended architecture would be appreciated.

Thank you.

**Reply by frtess** [reply]
Kind of in the same boat atm. I really like the Zero Trust Tunnel for it's simplicity and the level of security it provides. But trying to figure out the frigate notifications too. Have you figured it out yet ?

**Reply by vector-co-uz** [reply]
> Kind of in the same boat atm. I really like the Zero Trust Tunnel for it's simplicity and the level of security it provides. But trying to figure out the frigate notifications too. Have you figured it out yet ?

I don't know how it happened, but notifications work. Frigate version 0.17 beta.
